### ISSUE TITLE
Redshift: Automatically insert in sort key order in incremental models

### ISSUE THREAD

**Issue by Limess** [fishtown-analytics/dbt#2719]
### Describe the feature
If sort keys are defined in DBT, run an ORDER BY matching the sort key definition within the `INSERT INTO` statement in the incremental materialization.

This should then mean that no vacuums are necessary in future to keep the table sorted.

This matches the Redshift [recommendation](https://docs.aws.amazon.com/redshift/latest/dg/c_best-practices-sort-key-order.html).

### Describe alternatives you've considered
This follows a small discussion on Slack https://getdbt.slack.com/archives/CJARVS0RY/p1597827002012000

After which I considered writing a DBT macro which only adds an `ORDER BY` statement based on the `sort` config in `incremental` models and placing it at the end of each `incremental` model by hand, but this seems like an unnecessary implementation detail to keep track of.

### Additional context
Only applies to Redshift when using compound (_not interleaved_) sort keys.

This might result in some slowdown for large incremental models at DBT runtime but should increase query performance later and reduce maintenance overhead.

I don't have a lot of context about how much benefit, or overhead this would bring in each context, this is mainly based on AWS recommendations.

### Who will this benefit?
Anyone using Redshift incremental models with sort keys.

**Comment by Limess** [comment]
Reading further on this it seems that this optimisation will only help if using a `COPY` command, not an `INSERT INTO`: https://github.com/awslabs/amazon-redshift-utils/issues/218

Closing this as a result as it won't improve things.

**Comment by jtcohen6** [comment]
I was pretty excited while reading this! Too bad to hear it's specific to `copy`. Even so, we'd still need to `vacuum sort` _eventually_, so as to reshuffle the incrementally inserted data into the original table. (And I think running lots of `vacuum sort` operations as a post-hook is a bad idea for [WLM reasons](https://docs.aws.amazon.com/redshift/latest/dg/wlm-queue-hopping.html).)

It reminds me of our discovery 12-18 months ago that we could avoid the need for immediate reclustering on Snowflake if we created tables with `order by`, in cluster key order (https://github.com/fishtown-analytics/dbt/pull/1591).

Redshift is actually quite good in this regard: by specifying sort keys at time of table creation, there is no need to include an explicit `order by sort1, sort2`.

**Comment by Limess** [comment]
Hmm, so after talking about a separate problem on the intermix slack, there was a suggestion that this many still be beneficial after all:
https://intermixio.slack.com/archives/CME7JEUA0/p1598282273081600?thread_ts=1598281223.080000&cid=CME7JEUA0

> Is that still beneficial regardless of sort-keys then due to the zone-maps? ...

> Zone maps exist and are maintained on Redshift - it’s based on Postgres after all.

> Sortkeys essentially guarantee non-overlapping zone maps.

> So yes - in all cases inserting in some consistent order w.r.t the ‘most sorted’ columns will always help

so this may be worth a shot.

**Comment by jtcohen6** [comment]
Thanks for linking the thread, I just took a look. It's been a while since we went deep on Redshift query plans. This would be a performance improvement specific to incremental runs (`insert`), right? For the initial table creation, is it sufficient to declare the `sort` key?

If that's true, we may consider appending incremental runs with `order by {{ config.get('sort', []) | join(',')`.

What I'd _love_ to see, if you have the time and interest, is some experimental benchmarking :) nothing fancy, just enough to know the added complexity is worth it!
